feat(core): update schematic to migrate to explicit query timing

devversion opened the pull request:
Introduces an update schematic for the `@angular/core` package that
automatically migrates pre-V8 `ViewChild` and `ContentChild` queries to
the new explicit timing syntax. This is not required yet, but with Ivy, queries
will be "dynamic" by default. Therefore specifying an explicit query timing
ensures that developers can smoothly migrate to Ivy (once it's the default).

Read more about the explicit timing API here: https://github.com/angular/angular/pull/28810

kara reviewed (changes requested):
Thanks so much for doing this! A few thoughts below.

devversion commented on packages/core/schematics/migrations/static-queries/index_spec.ts:
```diff
@@ -0,0 +1,257 @@
+        
+        @Component({template: '<span #test></span>'})
+        export class MyComp {
+          @${queryType}('test', ) query: any;
```
That's actually by intention as I want to verify that we don't generate something like `@ViewQuery(a, {static: true}, )` which would be invalid. Will add a comment.

kara replied:
Ah, gotcha. A comment sounds good :+1

devversion commented on packages/core/schematics/migrations/static-queries/index_spec.ts:
```diff
@@ -0,0 +1,257 @@
+          .toContain(`@${queryType}('test', { static: true }) query: any;`);
+    });
+
+    it('should mark queries used in "ngAfterContentInit" as static', () => {
```
Makes sense. Will update that.

JoostK commented on packages/core/schematics/migrations/static-queries/index_spec.ts:
```diff
@@ -0,0 +1,257 @@
+            this.b();
+          }
+          
+          b() {
```
Could you add a testcase with recursion as well, you took care of it in the code 👍

devversion replied:
Good idea. will do that!

JoostK commented on packages/core/schematics/migrations/static-queries/migration.ts:
```diff
@@ -0,0 +1,97 @@
+    sourceFile: ts.SourceFile) {
+  const queryExpr = query.decorator.node.expression as ts.CallExpression;
+  const queryArguments = queryExpr.arguments;
+  const timingPropertyAssignment = ts.createPropertyAssignment(
```
Is it desirable to set `static: false` when that is the default behavior? Specifying it explicitly like this may indicate an explicit choice by a user, which is not true for automated migrations like this one.

kara replied:
Yes, we do want to explicitly set `static: false` here. The point of this schematic is to transition users on View Engine to the way queries work in Ivy (without actually having to update to Ivy). 

If we left this as `undefined`, View Engine would fall back to its default query sorting, so queries that happen to happen to have no results in embedded views would be resolved statically. This could be a problem if later the user adds behavior in `ngOnInit` that relies on this resolution timing (once switching to Ivy, it will break).

If we explicitly set this to `false`, it will be treated as dynamic in View Engine regardless of where the query results live. So it's guaranteed to work until switching to Ivy.

JoostK replied:
@kara thanks for the explanation, that makes sense!

JoostK commented on packages/core/schematics/migrations/static-queries/typescript/class_declaration.ts:
```diff
@@ -0,0 +1,32 @@
+  return node.heritageClauses.filter(clause => clause.token === ts.SyntaxKind.ExtendsKeyword)
+      .reduce((types, clause) => types.concat(clause.types), [] as ts.ExpressionWithTypeArguments[])
+      .map(typeExpression => typeExpression.expression)
+      .filter(expression => ts.isIdentifier(expression)) as ts.Identifier[];
```
nit; using the function reference instead of a new lambda will retain the `node is ts.Identifier` return type of `ts.isIdentifier(node: ts:Node)`, which in turn selects the `filter` overload that narrows the result type accordingly :)

```suggestion
      .filter(ts.isIdentifier);
```

JoostK commented on packages/core/schematics/utils/project_tsconfig_paths.ts:
```diff
@@ -0,0 +1,70 @@
+ * the CLI project could be still using `.angular-cli.json` instead of thew new config.
+ */
+function getWorkspaceConfigGracefully(tree: Tree): any {
+  const path = defaultWorkspaceConfigPaths.filter(filePath => tree.exists(filePath))[0];
```
```suggestion
  const path = defaultWorkspaceConfigPaths.find(filePath => tree.exists(filePath));
```

JoostK reviewed (commented):
Awesome work! 🎉

devversion committed:
feat(core): update schematic to migrate to explicit query timing

Introduces an update schematic for the "@angular/core" package
that automatically migrates pre-V8 "ViewChild" and "ContentChild"
queries to the new explicit timing syntax. This is not required
yet, but with Ivy, queries will be "dynamic" by default. Therefore
specifying an explicit query timing ensures that developers can
smoothly migrate to Ivy (once it's the default).

Read more about the explicit timing API here:
https://github.com/angular/angular/pull/28810

devversion commented:
@kara @JoostK I've addressed the feedback. thank you!

jelbourn commented on packages/core/schematics/utils/project_tsconfig_paths.ts:
```diff
@@ -0,0 +1,70 @@
+ * Gets all tsconfig paths from a CLI project by reading the workspace configuration
+ * and looking for common tsconfig locations.
+ */
+export function getProjectTsConfigPaths(tree: Tree): string[] {
```
Is there an issue filed for moving the functionality in this directory to the devkit directly? It now exists both in cdk and in angular itself.

devversion replied:
There is no real issue about this in particular, but there is a more generic issue that is about Angular workspaces (which includes logic to read the workspace config). https://github.com/angular/angular-cli/issues/12134

devversion replied:
Thinking about it more. The issue is more about Angular AST helpers. Though I'm not sure if we should create an issue for this in the devkit? Isn't such logic (like the `getProjectTsconfigPaths`) always specific to a given schematic?

I'm not sure if it's actually reasonable at all to expose something like that in the devkit? Let me know if you want me to create an issue for it.

kara commented on packages/core/schematics/migrations/static-queries/angular/analyze_query_usage.ts:
```diff
@@ -0,0 +1,69 @@
+  // In case there are no lifecycle hooks defined which could access a query
+  // statically, we can consider the query as dynamic as nothing in the class declaration
+  // could reasonably access the query in a static way.
+  if (!staticQueryHooks.length) {
```
What if the class has no hooks, but the query result is used in a setter for an input property? These properties are set before `onInit`, so they would need to be static.

devversion replied:
That sounds like another case we need to handle. This also means that we need to parse the directive/component metadata and also identify setters marked as inputs within the `inputs: [..]` array.

devversion replied:
I wanted to split things like handling the micro/macro tasks in follow-up's, but I think I'll just combine this into this PR. Marking as "wip".

aitboudad replied:
what's about a third-party library which relies on that (no hooks, no
setter) in our case we allow users to create a panel wrapper around a component dynamically.
```ts
@Component({
selector: 'formly-wrapper-panel',
template: `
 <div class="card">
   <ng-container #fieldComponent></ng-container>
 </div>
`,
})
export class PanelWrapperComponent extends FieldWrapper {
  @ViewChild('fieldComponent', {read: ViewContainerRef}) fieldComponent: ViewContainerRef;
}
```

devversion replied:
I thought about that, but wasn't able to come up with a way which allows us to handle that.

Since we can't find/analyze external projects which rely on the exposed query, the best thing is probably printing out which queries have changed and asking the developer to double-check these. Library creators need to be careful in particular.

<sup>We can show the errors only if the CLI project is set up as a "library" and not an application</sup>

devversion committed:
fixup! feat(core): update schematic to migrate to explicit query timing

Properly handle chained class inheritances

kara reviewed (approved):
LGTM for behavior. @alexeagle should probably look at bazel sections.

Note: Looks like linter is upset

devversion committed:
fixup! feat(core): update schematic to migrate to explicit query timing

Detect input setters with static query usage

alexeagle reviewed (approved):
no issues from Bazel

AndrewKushnir commented:
[Presubmit](http://test/OCL:236696593:BASE:236696647:1551729241051:1cfa6df1)

alexeagle commented on packages/core/schematics/migrations/static-queries/angular/analyze_query_usage.ts:
```diff
@@ -0,0 +1,83 @@
+  // access it statically. e.g.
+  //  (1) queries used in the "ngOnInit" lifecycle hook are static.
+  //  (2) inputs with setters can access queries statically.
+  const possibleStaticQueryNodes: ts.Node[] = classDecl.members.filter(m => {
```
using google3 TS settings, this fails to type-check
```
INFO: Found 1 target...
ERROR: /google/src/cloud/alexeagle/sync_28983/google3/third_party/javascript/angular2/rc/packages/core/schematics/BUILD:5:1: Compiling TypeScript (devmode) //third_party/javascript/angular2/rc/packages/core/schematics:schematics failed (Exit 1)
third_party/javascript/angular2/rc/packages/core/schematics/migrations/static-queries/angular/analyze_query_usage.ts:56:72 - error TS7030: Not all code paths return a value.

56   const possibleStaticQueryNodes: ts.Node[] = classDecl.members.filter(m => {
                                                                          ~~~~~~

Target //third_party/javascript/angular2/rc/packages/core/schematics:schematics failed to build
```

devversion replied:
Fixed.

alexeagle commented on packages/core/schematics/utils/project_tsconfig_paths_spec.ts:
```diff
@@ -0,0 +1,47 @@
+/**
```
we exclude specs from the g3 sync by following our convention of having them in a separate `/test/` folder.
how about moving these to `packages/core/schematics/test/*`

devversion replied:
Sure. Done!

alexeagle reviewed (commented):
caretaker: I was able to get this green in g3 by 
1) fixing the type-check error mentioned in a comment here
2) adding a new file `third_party/javascript/angular2/rc/packages/core/schematics/BUILD` containing
```
load("//javascript/typescript:build_defs.bzl", "ts_library")

licenses(["notice"])  # MIT License

ts_library(
    name = "schematics",
    srcs = glob(["migrations/**/*.ts"], exclude = [
        "migrations/static-queries/*.ts",
        "**/*_spec.ts",
    ]),
    deps = [
         "//third_party/javascript/node_modules/typescript:es2015.core",
         "//third_party/javascript/node_modules/typescript:typings",
    ]
)
```

note that this also has to be treated specially in copybara since this file would be g3-only.

The fundamental problem is that we don't share the BUILD files between github and g3. Would be an interesting project.

devversion committed:
fixup! feat(core): update schematic to migrate to explicit query timing

Fix no implicit returns G3 error. Move specs into packages/core/schematis/test/* folder

alexeagle commented:
Mailed cl/236891637 to @AndrewKushnir so that the new schematics/ directory here won't be compiled in g3 yet

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:236696593:BASE:236895398:1551816616838:87499377)

AndrewKushnir commented:
@alexeagle it looks like [TAP](https://test.corp.google.com/ui#id=OCL:236696593:BASE:236895398:1551816616838:87499377) and [Guitar](https://sponge.corp.google.com/invocation?id=06ebf83c-fc83-4859-ba7a-e1f4444eadc8) tests are still failing. Could you please help look into this?

AndrewKushnir commented:
[Presubmit #3](https://test.corp.google.com/ui#id=OCL:236696593:BASE:236911528:1551821840088:30ce9658) (after additional changes in g3)

AndrewKushnir commented:
Presubmit was successful, I've merged this PR. Thanks everyone for the help!
